[aws-ami-share] introduce new integration

maorfr opened the pull request:
part of https://issues.redhat.com/browse/APPSRE-4621

depends on https://github.com/app-sre/qontract-schemas/pull/91

design doc: https://gitlab.cee.redhat.com/service/app-interface/-/merge_requests/34669

this integration is an idea from @BumbleFeng and will help simplify this: https://github.com/app-sre/qontract-reconcile/pull/2100#discussion_r817030105

this PR introduces a new integration that shares AMIs between AWS accounts.
by defining a `sharing` section in an AWS account file, we turn on sharing, according to the `provider`.
we start out by only supporting the `ami` provider, which will share images and copy image tags.

still TODO: tests and type hints

Piojo commented on reconcile/utils/aws_api.py:
```diff
@@ -874,6 +883,32 @@ def get_vpcs_details(self, account, tags=None, route_tables=False):
 
         return results
 
+    def get_amis_details(self, account, owner_account, regex):
```
Can we have type hints here, please?

maorfr replied:
i'll add type hints, also wrote it in the description

Piojo commented on reconcile/aws_ami_share.py:
```diff
@@ -0,0 +1,74 @@
+MANAGED_TAG = {"Key": "managed_by_integration", "Value": QONTRACT_INTEGRATION}
+
+
+def filter_accounts(accounts):
```
Type hints, please. It took me a while to understand why you're not returning `sharing_accont_names` in the first place.

maorfr replied:
will do

jfchevrette commented on reconcile/queries.py:
```diff
@@ -309,12 +309,25 @@ def get_github_orgs():
       region
     }
     partition
+    {% if sharing %}
```
I don't understand why this templating is needed, can you explain? It seems like `get_aws_accounts` could simply always return the sharing data and then in the integration since you're testing if ` sharing` is falsy (you could also check that length is > ?) the code would end up the same

Perhaps the only reason I can see is to make all the other `get_aws_accounts` graphql queries a little bit smaller and more efficient by not requesting unnecessary data?

maorfr replied:
exactly. in the future, we may add more graphql objects under this section. this templating will make sure we do not need to update the allowed schemas list for other integrations.

jfchevrette reviewed (changes requested):
LGTM but marking as "changes requested" until tests and type hints are added

maorfr commented:
@Piojo i recently added tests for aws api: https://github.com/app-sre/qontract-reconcile/blob/master/reconcile/test/test_utils_aws_api.py

Piojo commented:
Woohoo! I missed that one!

steveteahan commented:
I think I understand the problem pretty well. I see that tests/types will be added. Should we also add this new integration to `README.md` and what is the plan for user-facing docs?

geoberle reviewed (approved):
LGTM

maorfr commented:
@steveteahan i've added a design doc and user facing docs in https://gitlab.cee.redhat.com/service/app-interface/-/merge_requests/34669. also added the command to readme.

lechuk47 commented:
AMIS are regional resources. They are  only usable in the same region they belong as described here[1]. Maybe we should ensure that the accounts have at least 1 common `supportedDeploymentRegion`. 

* This logic could be better placed in the app-interface pr_check. 

[1]: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/sharingamis-explicit.html

maorfr commented:
good point @lechuk47! added support for defining a region.
note: if the integration fails due to it, it fails app-interface pr-check.

geoberle commented:
> good point @lechuk47! added support for defining a region. note: if the integration fails due to it, it fails app-interface pr-check.
so this imply that the AMI must be build in each region it is required? or copied there so it can be shared?

maorfr commented:
> so this imply that the AMI must be build in each region it is required? or copied there so it can be shared?

indeed. but this is out of scope here IMO, as we only share between accounts, and not replicate across regions.

geoberle commented:
> indeed. but this is out of scope here IMO, as we only share between accounts, and not replicate across regions.

sounds ok for the time being. if the requirement of providing cross region would come up, we can deal with that since the region in the schema change is optional.

geoberle commented on reconcile/utils/aws_api.py:
```diff
@@ -889,7 +889,7 @@ def get_amis_details(self,
                          owner_account: dict[str, Any],
                          regex: str,
-                         region: Optional[str] = None) -> List[Dict[str, Sequence[Collection[str]]]]:
+                         region: Optional[str] = None) -> List[Dict[str, Any]]:
```
Any is a valid choice here in my opinion. the `item` constructed below has mixed type contents so automatic inference is hard. you could theoretically introduce a dataclass to represent the item but that would be a bit overengineered.

fishi0x01 commented on reconcile/aws_ami_share.py:
```diff
@@ -0,0 +1,86 @@
+MANAGED_TAG = {"Key": "managed_by_integration", "Value": QONTRACT_INTEGRATION}
+
+
+def filter_accounts(accounts: list[dict[str, Any]]):
```
According to style guide, abstract collection types should be used in favor of concrete collection types for method/function arguments.

```
def filter_accounts(accounts: Iterable[Mapping[str, Any]]) ->  list[dict[str, Any]]:
```

maorfr replied:
thanks!

maorfr replied:
updated in https://github.com/app-sre/qontract-reconcile/pull/2216/commits/e406c89616e1fa6f666b74e75f3815c0524c1b08

fishi0x01 commented on reconcile/aws_ami_share.py:
```diff
@@ -0,0 +1,86 @@
+
+
+def get_region(
+    share: dict[str, Any], src_account: dict[str, Any], dst_account: dict[str, Any]
```
```
def get_region(
    share: Mapping[str, Any], src_account: Mapping[str, Any], dst_account: Mapping[str, Any]
) -> str
```

maorfr replied:
thanks! https://github.com/app-sre/qontract-reconcile/pull/2216/commits/eae668d0c98f58cb0d5fe532616cd0c0a8bf8f48

BumbleFeng commented on reconcile/aws_ami_share.py:
```diff
@@ -0,0 +1,86 @@
+                    )
+                    if not dry_run:
+                        aws_api.create_tag(dst_account, dst_ami_id, MANAGED_TAG)
+                src_ami_tags = src_ami["tags"]
```
```suggestion
                src_ami_tags: Sequence[TagTypeDef] = src_ami["tags"]
```

maorfr replied:
went in a different direction. thanks!

BumbleFeng commented on reconcile/aws_ami_share.py:
```diff
@@ -0,0 +1,87 @@
+from mypy_boto3_ec2.type_defs import TagTypeDef
+
+QONTRACT_INTEGRATION = "aws-ami-share"
+MANAGED_TAG = TagTypeDef(Key="managed_by_integration", Value=QONTRACT_INTEGRATION)
```
```suggestion
if TYPE_CHECKING:
    from mypy_boto3_ec2.type_defs import TagTypeDef
else:
    TagTypeDef = object

MANAGED_TAG: TagTypeDef = {"Key": "managed_by_integration", "Value": QONTRACT_INTEGRATION}
```

maorfr replied:
went in a different direction. thanks!

fishi0x01 commented on reconcile/aws_ami_share.py:
```diff
@@ -0,0 +1,90 @@
+    aws_api = AWSApi(1, sharing_accounts, settings=settings)
+
+    for src_account in sharing_accounts:
+        sharing = src_account.get("sharing")
```
nit: sth I learned from you some weeks ago

```
sharing = src_account.get("sharing") or []
```

maorfr replied:
indeed. but i have learned in the past that it's more efficient to check `if sharing` and not to instantiate an empty list.

it's also more suitable here IMO because it allows us to add more code after the following loop without "missing" that there was no `sharing` section.

fishi0x01 commented on reconcile/utils/aws_api.py:
```diff
@@ -874,6 +893,47 @@ def get_vpcs_details(self, account, tags=None, route_tables=False):
+        launch_permission: LaunchPermissionModificationsTypeDef = {'Add': [{'UserId': share_account_uid}]}
+        image.modify_attribute(LaunchPermission=launch_permission)
+
+    def create_tag(self,
```
```
def create_tag(self, account: Mapping[str, Any], resource_id: str, tag: Mapping[str, str]):
```

maorfr replied:
thanks! https://github.com/app-sre/qontract-reconcile/pull/2216/commits/537ebe5c23a1c57fc7f9911237a801a7a840a926

fishi0x01 commented on reconcile/utils/aws_api.py:
```diff
@@ -874,6 +893,47 @@ def get_vpcs_details(self, account, tags=None, route_tables=False):
+        return self._filter_amis(images, regex)
+
+    def share_ami(self,
+                  account: dict[str, Any],
```
```
account: Mapping[str, Any]
```

maorfr replied:
thanks! https://github.com/app-sre/qontract-reconcile/pull/2216/commits/537ebe5c23a1c57fc7f9911237a801a7a840a926

Piojo reviewed (approved):
Minor nits, but it LVGTM.

maorfr commented on reconcile/utils/aws_api.py:
```diff
@@ -874,6 +893,47 @@ def get_vpcs_details(self, account, tags=None, route_tables=False):
         return results
 
+    @staticmethod
+    def _filter_amis(images: List[ImageTypeDef], regex: str) -> List[Dict[str, Any]]:
```
thanks! https://github.com/app-sre/qontract-reconcile/pull/2216/commits/537ebe5c23a1c57fc7f9911237a801a7a840a926
